### ISSUE TITLE
Report errors returned in JSON response

### ISSUE THREAD

**Issue by falbrechtskirchinger** [shrink/actions-docker-registry-tag#20]
As documented [here](https://docs.docker.com/registry/spec/api/#pushing-an-image-manifest) the server returns error messages in the JSON response body. These should be reported to the user in addition to the generic failure message (`<repository:tag> could not be tagged with <tag>`).

**Comment by shrink** [comment]
@falbrechtskirchinger hi! Thank you very much for the feedback. I agree the absence of error reporting in that error case is not ideal. I've created a Pull Request #21 to output the error message (when debugging is enabled) but I'm struggling to identify a scenario in which this error is expected to happen. Can you share the situation in which you encountered an error? I'll then be able to add a test and merge this. Thanks!

**Comment by falbrechtskirchinger** [comment]
Tagging failed because I forgot to add
```yaml
    permissions:
      packages: write
```
to my GitHub workflow.

The error returned was `installation not allowed to Write organization package`.

Thanks!

**Comment by shrink** [comment]
@falbrechtskirchinger perfect, thank you very much. I've added a test for that case.

Just for future reference, you don't need to add permissions within your Action configuration, rather, you can configure permissions via the package itself:

1. Visit the package, e.g: https://github.com/shrink/actions-docker-registry-tag/pkgs/container/actions-docker-registry-tag
2. Click "Package Settings" at the bottom right
3. Select the appropriate repositories under "Manage Actions access" and enable "Role: Write"

I often trip myself up on these permissions because they're not obvious upfront. The caveat with your proposal is that it will only work if the actor who triggers the action has permission to write to the package (essentially, "permissions" in the Action configuration is telling it to inherit the permissions of the actor). I hope someday GitHub will make the package permissions a little clearer!

**Comment by falbrechtskirchinger** [comment]
> Just for future reference, you don't need to add permissions within your Action configuration, rather, you can configure permissions via the package itself:

That's not particularly convenient for a large number of packages (hundreds). And even for just a handful, I prefer setting permissions in the workflow.

> The caveat with your proposal is that it will only work if the actor who triggers the action has permission to write to the package

This should be the case for a `push` event.

> My understanding is that if GitHub Actions create the package, it auto has write permissions

Maybe within the same job? I'll have to check that. The Docker images in question are created via one workflow and then tagged in another once they have been verified. I do set permissions in my build workflow and I believe that was copied from the `docker/build-push-action` example.

**Comment by shrink** [comment]
@falbrechtskirchinger  I haven't looked at this in quite a while, so I wouldn't bet my GitHub account on it, but _if memory serves_, the reason GitHub introduced the repository level package permissions was to remove the need for the use of Personal Access Tokens which in turn removed the dependency on pushes coming from users who had access to secrets and allowed more expressive permissions -- i.e: an actor can write packages when they're created via GitHub Actions based on a push, but I can't write packages by pushing to the registry using my GitHub credentials.

I'm a little low on time right now so I can't do it myself, but I think the test to verify the behaviour would be:

1. Create a new repository
2. Configure Actions to be restricted ("Read repository contents permissions")
3. Run a Workflow (without any explicit permissions) that pushes a package (for the first time, i.e: the package doesn't exist before the Workflow runs)
4. Navigate to the package settings
5. Confirm that the repository has defaulted to "Write" permissions

Then:

1. For the new package, change the role of the repository (under "Manage Actions access") to "Read"
2. Disable "Inherited access"
3. Modify the test Workflow to include the `permissions: packages: write` piece you shared above
4. Run the Workflow again

If my understanding is correct, the Workflow will fail, because you do not have permission to write (which was previously granted by Inherited Access) which your explicit `permissions` configuration cannot bypass. 

I'm happy to improve the documentation to make all this clearer; I think the current note...

> [...] Ensure the Actions token has permission to write to the package repository by granting access as described in ["About GitHub Packages with GitHub Actions"](https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#upgrading-a-workflow-that-accesses-ghcrio).[...]

...could be clearer!

**Comment by falbrechtskirchinger** [comment]
What point are you trying to make? I don't contest this behavior, so I'm unsure where we disagree.

My point, though, is that I don't care to change the package permissions from their default through the web interface or API, and I shouldn't have to. Inheriting permissions from the repository suits my use case.

Would you be amenable to an addition similar to the following (details being up for debate)?
```yaml
permissions:
  packages: write # may or may not be required depending on GitHub settings
  # contents: read # required to read the repository, i.e., actions/checkout
```

I don't understand why you reject adding the `permissions` key. What is your concern? At worst, it's redundant. Am I missing something?

Just trying to spare people from losing a few hours to this issue, although, thanks to #21 that time should be much shorter now (most of the time was spent trying to figure out how to deploy a fork of this action).

> I'm happy to improve the documentation to make all this clearer; I think the current note...
> 
> > [...] Ensure the Actions token has permission to write to the package repository by granting access as described in ["About GitHub Packages with GitHub Actions"](https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#upgrading-a-workflow-that-accesses-ghcrio).[...]
> 
> ...could be clearer!

Making it more prominent would be great as I read past it entirely. Adding a dedicated "Permissions" section, instead of sneaking it in with "Inputs," should help.

**Comment by falbrechtskirchinger** [comment]
Another option:
```yaml
# may be required depending on GitHub settings
# permissions:
#   packages: write
#   contents: read # required to read the repository, i.e., actions/checkout
```

Having some reference in the example is important IMHO.

**Comment by shrink** [comment]
Sorry about the lack of clarity. My point is that when a package is not configured to grant GitHub Actions write permissions (which usually happens automatically but sometimes does not) there are some workarounds (such as the one you've proposed) but they do not resolve the underlying issue and will not work for everyone. For example, your workaround depends upon the GitHub Action's actor (i.e: the authentication context within which the Workflow executes) having permissions that GitHub Actions does not, which may not always be the case.

I am very happy to accept documentation improvements and certainly agree that if you encountered challenges in using this action because the documentation was unclear, then the documentation has room for improvement, _however_ I want to ensure that accurate information is provided.

My specific concern with the workaround you've used is that it shifts the problem further down the line; that is, by working on the GitHub Workflows, you have permissions that others who interact with a repository may not have, and so the workaround may work for you now, but may not work when someone else tries to interact with the repository -- and that's a much more difficult problem to track down.

I outlined my understanding of GitHub Action's permissions to try and provide some additional context, and to ensure that we're on the same page, so that if I am wrong, then you're able to identify exactly where I am wrong, and we can come to an accurate understanding together (certainly possible I am wrong, I haven't been deep into GitHub Actions recently and there have been several great improvements).

So, in summary, I will gratefully accept documentation improvements that better surface the information about the required package configuration **and/or** if you can demonstrate that my concerns about permissions are unfounded (and so your workaround is a suitable alternative to configuring the package to be writable by GitHub Actions) then I will gratefully accept the improvements to the examples you've shared.

**Comment by falbrechtskirchinger** [comment]
Thank you for your in-depth explanation.

I disagree with your characterization of *workaround*. GitHub offers one high-level switch to set permissions on a repository level. The options are "permissive" or "restrictive." The setting controls **default** permissions for workflows. Workflows can request additional permissions within the bounding set of permissions granted to the actor.

Of course, the actor needs to have write permissions in the first place, but the effective permissions in the workflow can still be lower due to the "restrictive" setting.

> My specific concern with the workaround you've used is that it shifts the problem further down the line; that is, by working on the GitHub Workflows, you have permissions that others who interact with a repository may not have,

What kind of interactions do you have in mind? I'm only considering the owner pushing a git tag, triggering a workflow run that adds corresponding tags to an image.
In my view, I'm not shifting anything; the actor **has** permissions (as a precondition), but the workflow may not (depending on the aforementioned repository setting). I'm just raising the workflow permissions to the actor's.

> and so the workaround may work for you now, but may not work when someone else tries to interact with the repository -- and that's a much more difficult problem to track down.

If you're thinking of forks, that's not a concern for me because I'm using `with: repository: ${{ github.repository_owner }}/my-docker-image`.
Because I'm working on a PR and am operating on a fork, this was one of the first changes I made to the existing workflow.

(It occurs to me that when we say "repository," we may be referring to different things. I'm almost always talking about git/GitHub repositories. In fact, the input name "repository" confused me a little before I brushed up on Docker terminology. Shows how unfamiliar I am with Docker. ;-))

Outside of that, I'm not sure what other interactions you're talking about. Adding tags is a privileged operation, as far as I'm concerned. No one else is supposed to interact with my (GHCR) repository in ways that require anything but read permissions. :man_shrugging:

In any case, inline comments can clarify the circumstances when `permissions:` may be needed.

**Comment by falbrechtskirchinger** [comment]
Alright, I don't think the default package permissions are an incorrect configuration, but I do understand where you're coming from. Ultimately, I'd like to thank you for adding the debug output. That should help the next person out a lot. I was able to resolve the issue immediately after reading the error response.

I'll take you up on improving the documentation over the weekend. I'll submit a quick draft now so I have something to remind me.
